### DISCUSSION TITLE
Missing I2c settings web page in custom firmware

### DISCUSSION THREAD

**Discussion by stefanoesmeris** [arendst/Tasmota#13457] [Q&A] [answered]
### PROBLEM DESCRIPTION
I have an I2C issue that is driving me crazy. I've searched the web and read documents and issues for hours and still can't figure out what's wrong. In short: I customized and compiled tasmota (the new version 10) to include I2c PCF8574 support.
After loading the firmware and configuring the SDA and SCL pins, it even recognizes the device at the correct address with the i2cscan command, but the specific web page for this driver is not displayed as described in the documentation.

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [X] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [X] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [X] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [X] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [ ] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [X] Device used (e.g., Sonoff Basic): Generic
- [X] Tasmota binary firmware version number used: 10
  - [ ] Pre-compiled
  - [X] Self-compiled
- [X] Flashing tools used: ESP8266Flasher
- [X] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:
23:00:55.621 CMD: Backlog Template; Module; GPIO 255
23:00:55.687 RSL: RESULT = {"NAME":"Generic","GPIO":[1,1,1,1,1,1,1,1,1,1,1,1,1,1],"FLAG":0,"BASE":18}
23:00:55.892 RSL: RESULT = {"Module":{"0":"Generic"}}
23:00:56.147 RSL: RESULT = {"GPIO0":{"640":"I2C SDA"},"GPIO1":{"0":"Nenhum"},"GPIO2":{"608":"I2C SCL"},"GPIO3":{"0":"Nenhum"},"GPIO4":{"0":"Nenhum"},"GPIO5":{"0":"Nenhum"},"GPIO9":{"0":"Nenhum"},"GPIO10":{"0":"Nenhum"},"GPIO12":{"0":"Nenhum"},"GPIO13":{"0":"Nenhum"},"GPIO14":{"0":"Nenhum"},"GPIO15":{"0":"Nenhum"},"GPIO16":{"0":"Nenhum"},"GPIO17":{"0":"Nenhum"}


```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:
not using rules


```
- [ ] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:
23:03:00.070 CMD: Status 0
23:03:00.075 RSL: STATUS = {"Status":{"Module":0,"DeviceName":"Tasmota","FriendlyName":["Tasmota"],"Topic":"tasmota_0E837A","ButtonTopic":"0","Power":0,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0}}
23:03:00.081 RSL: STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota/release/tasmota.bin.gz","RestartReason":"Software/System restart","Uptime":"0T00:30:34","StartupUTC":"2021-10-24T01:32:26","Sleep":50,"CfgHolder":4617,"BootCount":16,"BCResetTime":"2021-10-23T21:28:22","SaveCount":30,"SaveAddress":"F6000"}}
23:03:00.085 RSL: STATUS2 = {"StatusFWR":{"Version":"10.0.0.1(tasmota)","BuildDateTime":"2021-10-23T21:51:07","Boot":6,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8266EX","CR":"369/699"}}
23:03:00.089 RSL: STATUS3 = {"StatusLOG":{"SerialLog":2,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["Stefano & Rosinacia",""],"TelePeriod":300,"Resolution":"558180C0","SetOption":["00008009","2805C8000100060000005A0A000000000000","00000080","00006000","00004000"]}}
23:03:00.098 RSL: STATUS4 = {"StatusMEM":{"ProgramSize":614,"Free":388,"Heap":26,"ProgramFlashSize":4096,"FlashSize":4096,"FlashChipId":"1640EF","FlashFrequency":40,"FlashMode":2,"Features":["00000416","8FDAC787","04368001","000000CF","000013C0","C000FB81","00004004","00001000","00000020"],"Drivers":"1,2,3,4,5,6,7,8,9,10,12,16,18,19,20,21,22,24,26,27,28,29,30,35,37,45","Sensors":"1,2,3,4,5,6"}}
23:03:00.103 RSL: STATUS5 = {"StatusNET":{"Hostname":"tasmota-0E837A-0890","IPAddress":"192.168.145.43","Gateway":"192.168.145.254","Subnetmask":"255.255.255.0","DNSServer1":"192.168.145.254","DNSServer2":"9.9.9.9","Mac":"2C:3A:E8:0E:83:7A","Webserver":2,"HTTP_API":1,"WifiConfig":4,"WifiPower":17.0}}
23:03:00.107 RSL: STATUS6 = {"StatusMQT":{"MqttHost":"","MqttPort":1883,"MqttClientMask":"DVES_%06X","MqttClient":"DVES_0E837A","MqttUser":"DVES_USER","MqttCount":0,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}}
23:03:00.112 RSL: STATUS7 = {"StatusTIM":{"UTC":"2021-10-24T02:03:00","Local":"2021-10-23T23:03:00","StartDST":"2021-03-28T02:00:00","EndDST":"2021-10-31T03:00:00","Timezone":"-03:00","Sunrise":"05:37","Sunset":"18:38"}}
23:03:00.116 RSL: STATUS10 = {"StatusSNS":{"Time":"2021-10-23T23:03:00"}}
23:03:00.121 RSL: STATUS11 = {"StatusSTS":{"Time":"2021-10-23T23:03:00","Uptime":"0T00:30:34","UptimeSec":1834,"Heap":26,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Wifi":{"AP":1,"SSId":"Stefano & Rosinacia","BSSId":"F4:92:BF:CD:5D:65","Channel":10,"Mode":"11n","RSSI":100,"Signal":-46,"LinkCount":1,"Downtime":"0T00:00:03"}}}

```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:


```

### TO REPRODUCE
_Steps to reproduce the behavior:_

user_config_override.h
#define USE_I2C                                // Add support for I2C
#define USE_PCF8574                          
// #define USE_PCF8574_SENSOR                   // enable PCF8574 inputs and outputs in SENSOR message
//#define USE_PCF8574_DISPLAYINPUT             // enable PCF8574 inputs display in Web page
// #define USE_PCF8574_MQTTINPUT                // enable MQTT message & rule process on input change detection : stat/%topic%/PCF8574_INP = {"Time":"2021-03-07T16:19:23+01:00","PCF8574-1_INP":{"D1":1}}

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by barbudor** [comment, marked as answer]
If your user_config_override.h is the above, you have still explicitly disabled the display on input pins as sensors in the web GUI by not removing the comment in front of '#define USE_PCF8574_DISPLAYINPUT'
Outputs should automatically appears as new POWER with their own on/off button

**Reply by stefanoesmeris** [reply]
Barbudor, I had understood that this option would only be if I intended to use it as an input. In my case, I want to use only as outputs for controlling relays. I've tried activating all the options in the user_config_override.h file and so far the configuration page has not been presented. Do you have anything else I can do?

**Reply by barbudor** [reply]
Oh sorry, you mean the configuration page to choose for inputs or outputs?

I can't check right now as I'm not at home.
I see a few possibilities
1) a bug
2) the page is no more in the "configuration" menu but have been moved to "consoles"
3) the PCF driver has not recognized the chip

Let's rule out 3) at first
- What is the result of 'i2cscan'? Is a i2c device detected at the expected address?
- what is the result of 'i2cdriver'? Number 22 should be listed without a '!' in front. If 22 is not listed something is wrong with your compile as the driver is not included. If you see '!22', driver is disabled and must be enabled by 'I2Cdriver22 1' as per docs.
- if all above are OK and still not working, set 'Weblog 3' and restart Tasmota. Look in the logs if you can see any message from I2C or PCF driver

**Reply by barbudor** [reply]
If the driver loads and you can't still find the menu entry for the page, you should be able to reach it by adding '/pcf' to the browser url

**Reply by sfromis** [reply]
Sorry, my comment about missing `USE_PCF8574` was wrong, there was a small slip up in the tool I used to decode, using wrong JSON instead of yours. Removed the text.

**Reply by stefanoesmeris** [reply]
_"do you mean the configuration page to choose inputs or outputs?"_ Yes. that's right

_"What is the result of 'i2cscan'?"_   RESULT = {"I2CScan":"Device(s) found at 0x27"} - What was expected.

 _"Is a i2c device detected at the expected address?"_ Yes. The address is 0x27

"_what is the result of 'i2cdriver'?_"  RESULT = {"I2CDriver":"2,7,22"} 
About this, the documentation at _https: //tasmota.github.io/docs/I2CDEVICES/_    USE_PCF8574 == 2 and USE_MCP230xx == 22  
I want to use the first one, PCF8574 and I thought it was enough to work.
But, I recompiled and added MPC230x support as well, to see if it worked.

Important Points:
The i2cscan only reach device, after setuped GPIO 0 to SDA and GPIO 2 to SCL 
On configuration -> Templates
Name Generic, Generic(18)
GPIO 0 -> SDA
GPIO 2 -> SCL

**Reply by barbudor** [reply]
> "_what is the result of 'i2cdriver'?_" RESULT = {"I2CDriver":"2,7,22"} About this, the documentation at _https: //tasmota.github.io/docs/I2CDEVICES/_ USE_PCF8574 == 2 and USE_MCP230xx == 22 I want to use the first one, PCF8574 and I thought it was enough to work. But, I recompiled and added MPC230x support as well, to see if it worked.

You are right, PCF is 2 and MCP is 22. There is a bad copy paste in the doc page for [PCF](https://tasmota.github.io/docs/PCF8574/#tasmota-configuration) which wrongly suggest 22.
 
> Important Points: The i2cscan only reach device, after setuped GPIO 0 to SDA and GPIO 2 to SCL On configuration -> Templates Name Generic, Generic(18) GPIO 0 -> SDA GPIO 2 -> SCL

Yes, you need I2C pins to be enabled for I2C commands to work.

So still no change ? No PCF page in the "Configure" menu ? (it would not be in the main page but in "Configure")
With `Weblog 3` and restart, what are the logs ?

**Reply by barbudor** [reply]
You don't need MCP, so may be better to disable it with `i2cdriver22 0`

**Reply by stefanoesmeris** [reply]
No changes. . .

The PCF page not found:
Not Found 404
URI: /pcf
Method: GET
Arguments: 0

#########################

Here you are the logs:
00:00:00.001 HDW: ESP8266EX
00:00:00.049 CFG: Lido da flash em F8, Contagem 28
00:00:00.054 QPC: Count 1
00:00:00.057 CFG: CR 370/699, Busy 0
00:00:00.060 ROT: Mode 1
00:00:00.061 PCF: Address 0x20 reserved for MCP320xx skipped
00:00:00.063 SRC: Restart
00:00:00.065 Projeto tasmota - Tasmota Versão 10.0.0.1(tasmota)-2_7_4_9(2021-10-24T11:48:21)
00:00:00.170 WIF: Verificando ligação...
00:00:00.170 WIF: Ligando...
00:00:00.507 WIF: Ligando ao PA1 Stefano & Rosinacia Channel 10 BSSId F4:92:BF:CD:5D:65 em modo 11n como tasmota-0E837A-0890...
00:00:01.752 WIF: Verificando ligação...
00:00:01.753 WIF: Ligado
00:00:02.004 HTP: Servidor WEB ativo em tasmota-0E837A-0890 com o endereço IP 192.168.145.42
00:00:02.452 NTP: Synch time...
00:00:04.712 RTC: UTC 2021-10-24T16:40:41, DST 2021-03-28T02:00:00, STD 2021-10-31T03:00:00
17:40:41.000 NTP: Synched
17:40:42.219 RSL: INFO1 = {"Info1":{"Module":"Generic","Version":"10.0.0.1(tasmota)","FallbackTopic":"cmnd/DVES_0E837A_fb/","GroupTopic":"cmnd/tasmotas/"}}
17:40:42.223 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota-0E837A-0890","IPAddress":"192.168.145.42"}}
17:40:42.234 RSL: INFO3 = {"Info3":{"RestartReason":"Software/System restart"}}
17:40:44.125 HTP: Menu principal
17:40:44.191 HTP: Menu principal
17:40:44.457 QPC: Reset
17:40:46.421 APP: Contagem de inicialização 18
17:40:46.426 RSL: STATE = {"Time":"2021-10-24T17:40:46","Uptime":"0T00:00:09","UptimeSec":9,"Heap":28,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Wifi":{"AP":1,"SSId":"Stefano & Rosinacia","BSSId":"F4:92:BF:CD:5D:65","Channel":10,"Mode":"11n","RSSI":100,"Signal":-46,"LinkCount":1,"Downtime":"0T00:00:03"}}
17:40:46.774 CFG: Salvo na flash em F7, Contagem 29, Bytes 4096
17:40:46.980 HTP: Console

**Reply by stefanoesmeris** [reply]
17:12:26.477 CMD: Status 0
17:12:26.479 SRC: WebConsole from 192.168.145.57
17:12:26.481 CMD: Grp 0, Cmnd 'STATUS', Idx 1, Len 1, Data '0'
17:12:26.485 RSL: STATUS = {"Status":{"Module":0,"DeviceName":"Tasmota","FriendlyName":["Tasmota"],"Topic":"tasmota_0E837A","ButtonTopic":"0","Power":0,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0}}
17:12:26.516 RSL: STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota/release/tasmota.bin.gz","RestartReason":"External System","Uptime":"0T00:04:28","StartupUTC":"2021-10-24T16:07:58","Sleep":50,"CfgHolder":4617,"BootCount":17,"BCResetTime":"2021-10-24T16:16:51","SaveCount":28,"SaveAddress":"F8000"}}
17:12:26.548 RSL: STATUS2 = {"StatusFWR":{"Version":"10.0.0.1(tasmota)","BuildDateTime":"2021-10-24T11:48:21","Boot":6,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8266EX","CR":"370/699"}}
17:12:26.567 RSL: STATUS3 = {"StatusLOG":{"SerialLog":4,"WebLog":4,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["Stefano & Rosinacia",""],"TelePeriod":300,"Resolution":"558180C0","SetOption":["00008009","2805C8000100060000005A0A000000000000","00000080","00006000","00004000"]}}
17:12:26.598 RSL: STATUS4 = {"StatusMEM":{"ProgramSize":623,"Free":380,"Heap":27,"ProgramFlashSize":4096,"FlashSize":4096,"FlashChipId":"1640EF","FlashFrequency":40,"FlashMode":2,"Features":["00000416","8FDAC787","04368001","000000CF","010013F1","C000FB81","00004004","00001000","00000020"],"Drivers":"1,2,3,4,5,6,7,8,9,10,12,16,18,19,20,21,22,24,26,27,28,29,30,35,37,45","Sensors":"1,2,3,4,5,6,29"}}
17:12:26.630 RSL: STATUS5 = {"StatusNET":{"Hostname":"tasmota-0E837A-0890","IPAddress":"192.168.145.42","Gateway":"192.168.145.254","Subnetmask":"255.255.255.0","DNSServer1":"192.168.145.254","DNSServer2":"9.9.9.9","Mac":"2C:3A:E8:0E:83:7A","Webserver":2,"HTTP_API":1,"WifiConfig":4,"WifiPower":17.0}}
17:12:26.655 RSL: STATUS6 = {"StatusMQT":{"MqttHost":"","MqttPort":1883,"MqttClientMask":"DVES_%06X","MqttClient":"DVES_0E837A","MqttUser":"DVES_USER","MqttCount":0,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}}
17:12:26.676 RSL: STATUS7 = {"StatusTIM":{"UTC":"2021-10-24T16:12:26","Local":"2021-10-24T17:12:26","StartDST":"2021-03-28T02:00:00","EndDST":"2021-10-31T03:00:00","Timezone":"+01:00","Sunrise":"07:25","Sunset":"17:43"}}
17:12:26.694 RSL: STATUS10 = {"StatusSNS":{"Time":"2021-10-24T17:12:26"}}
17:12:26.700 RSL: STATUS11 = {"StatusSTS":{"Time":"2021-10-24T17:12:26","Uptime":"0T00:04:28","UptimeSec":268,"Heap":27,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Wifi":{"AP":1,"SSId":"Stefano & Rosinacia","BSSId":"F4:92:BF:CD:5D:65","Channel":10,"Mode":"11n","RSSI":98,"Signal":-51,"LinkCount":1,"Downtime":"0T00:00:03"}}}

**Comment by barbudor** [comment]
From the driver's code:
`// Support I2C addresses 0x20 to 0x26 and 0x39 to 0x3F`

So 0x27 is not support because it is reserved for another sensor which doesn't have the flexibility to change the address
Also 0x20 is generally reserved for MCP23017
So can you please set your PCF to any other address between 0x21 and 0x26 ?

Sorry I forgot that. We should make it clear in the docs.

EDIT: It is already in the docs : https://tasmota.github.io/docs/PCF8574/#supported-i2c-addresses-and-number-of-pcf8574

**Reply by stefanoesmeris** [reply]
Yes, You are right.
I did change in driver PCF8574 source code, from 0x20 to 0x27, because my hardware is not flexible.
After this, all worked fine.
Thanks a lot!
